Unittest unexpected success should be translated to XPASS

pytest version: 5.4.x & current master, Python version: 3.8

## xfail background

pytest's unittest plugin tries to translate unittest concepts to pytest concept. One concept in pytest is xfail/XFAIL/XPASS (see [docs](https://docs.pytest.org/en/latest/skipping.html#xfail-mark-test-functions-as-expected-to-fail)).

Example:

```py
import pytest

@pytest.mark.xfail
def test_expected_failure():
    assert False

@pytest.mark.xfail
def test_unexpected_success():
    assert True
```

running:

```
x.py::test_expected_failure XFAIL
x.py::test_unexpected_success XPASS
```

also, if running with strict xfail, this becomes

```
x.py::test_expected_failure XFAIL
x.py::test_unexpected_success FAILED

============== FAILURES =============
______ test_unexpected_success ______
[XPASS(strict)] 
```

## unittest background

unittest also has expected-failure/unexpected-success support:

```py
import unittest

class ExpectedFailureTestCase(unittest.TestCase):
    @unittest.expectedFailure
    def test_expected_failure(self):
        self.assertTrue(False)

    @unittest.expectedFailure
    def test_unexpected_success(self):
        self.assertTrue(True)
```

running:

```
xu
----------------------------------------------------------------------
Ran 2 tests in 0.000s

FAILED (expected failures=1, unexpected successes=1)
ran@ran:~/src/pytest$ py -m unittest -v ut.py 
test_expected_failure (ut.ExpectedFailureTestCase) ... expected failure
test_unexpected_success (ut.ExpectedFailureTestCase) ... unexpected success

----------------------------------------------------------------------
Ran 2 tests in 0.000s

FAILED (expected failures=1, unexpected successes=1)
```

## pytest unittest plugin

Given the correspondence described above, when running the unittest under pytest I would expect `test_expected_failure` to be translated to an XFAIL -- this happens, and `test_unexpected_success` to be translated to an XPASS -- this doesn't happen:

```
ut.py::ExpectedFailureTestCase::test_expected_failure XFAIL                     [ 50%]
ut.py::ExpectedFailureTestCase::test_unexpected_success FAILED                  [100%]

====================================== FAILURES =======================================
___________________ ExpectedFailureTestCase.test_unexpected_success ___________________
Unexpected success
=============================== short test summary info ===============================
FAILED ut.py::ExpectedFailureTestCase::test_unexpected_success
============================ 1 failed, 1 xfailed in 0.13s =============================
```

To preserve the unittest behavior of unexpected-success being considered a failure, pytest should probably translate to xfail `strict=True`.

## Implementation notes

The current behavior is implemented in the skipping and unittest plugins using a `unexpectedsuccess_key` hack. Ideally this hack would be removed in favor of using straight xfail, this would also be a good code cleanup. Not sure how hard that would be to achieve.

--- a/src/_pytest/skipping.py
+++ b/src/_pytest/skipping.py
@@ -234,7 +234,6 @@ def evaluate_xfail_marks(item: Item) -> Optional[Xfail]:
 skipped_by_mark_key = StoreKey[bool]()
 # Saves the xfail mark evaluation. Can be refreshed during call if None.
 xfailed_key = StoreKey[Optional[Xfail]]()
-unexpectedsuccess_key = StoreKey[str]()
 
 
 @hookimpl(tryfirst=True)
@@ -271,15 +270,7 @@ def pytest_runtest_makereport(item: Item, call: CallInfo[None]):
     outcome = yield
     rep = outcome.get_result()
     xfailed = item._store.get(xfailed_key, None)
-    # unittest special case, see setting of unexpectedsuccess_key
-    if unexpectedsuccess_key in item._store and rep.when == "call":
-        reason = item._store[unexpectedsuccess_key]
-        if reason:
-            rep.longrepr = f"Unexpected success: {reason}"
-        else:
-            rep.longrepr = "Unexpected success"
-        rep.outcome = "failed"
-    elif item.config.option.runxfail:
+    if item.config.option.runxfail:
         pass  # don't interfere
     elif call.excinfo and isinstance(call.excinfo.value, xfail.Exception):
         assert call.excinfo.value.msg is not None

--- a/src/_pytest/unittest.py
+++ b/src/_pytest/unittest.py
@@ -30,10 +30,10 @@ from _pytest.python import Function
 from _pytest.python import PyCollector
 from _pytest.runner import CallInfo
 from _pytest.skipping import skipped_by_mark_key
-from _pytest.skipping import unexpectedsuccess_key
 
 if TYPE_CHECKING:
     import unittest
+    import twisted.trial.unittest
 
     from _pytest.fixtures import _Scope
 
@@ -273,24 +273,24 @@ class TestCaseFunction(Function):
             self._addexcinfo(sys.exc_info())
 
     def addUnexpectedSuccess(
-        self, testcase: "unittest.TestCase", reason: str = ""
+        self,
+        testcase: "unittest.TestCase",
+        reason: Optional["twisted.trial.unittest.Todo"] = None,
     ) -> None:
-        self._store[unexpectedsuccess_key] = reason
+        msg = "Unexpected success"
+        if reason:
+            msg += f": {reason.reason}"
+        # Preserve unittest behaviour - fail the test
+        try:
+            fail(msg, pytrace=False)
+        except fail.Exception:
+            self._addexcinfo(sys.exc_info())
 
     def addSuccess(self, testcase: "unittest.TestCase") -> None:
         pass
 
     def stopTest(self, testcase: "unittest.TestCase") -> None:
         pass
-
-    def _expecting_failure(self, test_method) -> bool:
-        """Return True if the given unittest method (or the entire class) is marked
-        with @expectedFailure."""
-        expecting_failure_method = getattr(
-            test_method, "__unittest_expecting_failure__", False
-        )
-        expecting_failure_class = getattr(self, "__unittest_expecting_failure__", False)
-        return bool(expecting_failure_class or expecting_failure_method)
 
     def runtest(self) -> None:
         from _pytest.debugging import maybe_wrap_pytest_function_for_tracing

--- a/src/_pytest/unittest.py
+++ b/src/_pytest/unittest.py
@@ -280,7 +280,7 @@ class TestCaseFunction(Function):
         msg = "Unexpected success"
         if reason:
             msg += f": {reason.reason}"
-        # Preserve unittest behaviour - fail the test
+        # Preserve unittest behaviour - fail the test. Explicitly not an XPASS.
         try:
             fail(msg, pytrace=False)
         except fail.Exception:

--- a/testing/test_unittest.py
+++ b/testing/test_unittest.py
@@ -765,7 +765,8 @@ def test_unittest_expected_failure_for_failing_test_is_xfail(
 
 @pytest.mark.parametrize("runner", ["pytest", "unittest"])
 def test_unittest_expected_failure_for_passing_test_is_fail(
-    pytester: Pytester, runner
+    pytester: Pytester,
+    runner: str,
 ) -> None:
     script = pytester.makepyfile(
         """
@@ -782,7 +783,11 @@ def test_unittest_expected_failure_for_passing_test_is_fail(
     if runner == "pytest":
         result = pytester.runpytest("-rxX")
         result.stdout.fnmatch_lines(
-            ["*MyTestCase*test_passing_test_is_fail*", "*1 failed*"]
+            [
+                "*MyTestCase*test_passing_test_is_fail*",
+                "Unexpected success",
+                "*1 failed*",
+            ]
         )
     else:
         result = pytester.runpython(script)

